### ISSUE TITLE
Minor error and question

### ISSUE THREAD

**Issue by halfluke** [dolevf/Damn-Vulnerable-GraphQL-Application#13]
In the Deep Recursion Query Attack solution, I think that "Title" is a typo: it triggers an error.

Apart from that, I have been trying to understand how one can come up to build such a circular query, and why a structure like owner -> paste -> owner -> paste -> owner....  does not work - and we need to add edges and node.
I've retrieved the Introspection schema with InQL and I've also fed it to GraphQL voyager, but still I'm not sure how I could figure out the correct structure of the circular query myself.
I clearly miss some Graphql concept here, I've tried to read about connections edges and nodes but I can't put everything together.
A few words of explanation or a reference to something to read would be really welcome.
T.I.A.

**Comment by dolevf** [comment]
hi there halfluke,

If you refer to `core/models.py` in the source code, you can see that we have an Owner class and a Paste class, and they each refer to the other (Owners have pastes, and pastes have owners). Owners can also have multiple pastes. In GraphQL, owners and pastes are nodes, and edges represent the relationship.

I actually think that Ruby's GraphQL documentation better explains this than Python's graphene.

https://graphql-ruby.org/pagination/connection_concepts.html

When it comes to security testing, if introspection is disabled, you will have to take notes while you dynamically test to identify these relationships. Some tools may be able to construct queries for you after some initial brute forcing (such as https://github.com/szski/shapeshifter and https://github.com/nikitastupin/clairvoyance)

Regarding your typo report - fixed it, thanks so much!

**Comment by dolevf** [comment]
Will be closing this tomorrow EOD if no other questions or concerns.

**Comment by halfluke** [comment]
Thank you for the info. I tried clairvoyance but it doesn't seem to work very well...

My concern is related to the ability to see a possible circular reference even when introspection is enabled.
In the end I was able to see it in DVGA, in GraphQL Voyager after unchecking "skip relay".
![image](https://user-images.githubusercontent.com/33841708/110503617-47ce2580-80f4-11eb-8ba4-ca37aeb99b35.png)

I am not aware of any tool that automatically looks for circular reference. In a real-world app (like the one I'm testing now) I get completely lost trying to determine if this vulnerability may exist (in order to test if a protection is implemented, we need to find a circular query in a complex schema and the correct syntax to call it...)
